db/user_js_runner.go

```diff
@@ -0,0 +1,384 @@
+		defer func() {
+			runner.currentDB = nil
+		}()
+		if err != nil {
+			base.ErrorfCtx(context.Background(), base.KeyJavascript.String()+": %s %s failed: %#v", runner.kind, runner.name, err)
+			return nil, runner.convertError(err)
+		}
+		return jsResult.Export()
+	}
+
+	return runner, nil
+}
+
+// Calls a javaScriptRunner's JavaScript function.
+func (runner *userJSRunner) CallWithDB(db *Database, mutationAllowed bool, args ...interface{}) (result interface{}, err error) {
+	runner.currentDB = db
+	runner.mutationAllowed = mutationAllowed
+	ctx := db.Ctx
+	var timeout time.Duration
+	if ctx != nil {
+		if deadline, exists := ctx.Deadline(); exists {
+			timeout = time.Until(deadline)
+			if timeout <= 0 {
+				return nil, sgbucket.ErrJSTimeout
+			}
+		}
+	}
+	runner.SetTimeout(timeout)
+	return runner.Call(args...)
+}
+
+// JavaScript error returned by a userJavaScriptRunner
+type jsError struct {
+	err        error
+	runnerKind string
+	runnerName string
+}
+
+func (jserr *jsError) Error() string {
+	return fmt.Sprintf("%v (while calling %s %q)", jserr.err, jserr.runnerKind, jserr.runnerName)
+}
+
+func (jserr *jsError) Unwrap() error {
+	return jserr.err
+}
+
+var HttpErrRE = regexp.MustCompile(`^HTTP:\s*(\d+)\s+(.*)`)
+
+func (runner *userJSRunner) convertError(err error) error {
+	if err == sgbucket.ErrJSTimeout {
+		return base.HTTPErrorf(408, "Timeout in JavaScript")
+	}
+	// Unfortunately there is no API on otto.Error to get the name & message separately.
+	// Instead, look for the name as a prefix. (See the `ottoResult` function below)
+	str := err.Error()
+	if strings.HasPrefix(str, "HTTP:") {
+		m := HttpErrRE.FindStringSubmatch(str)
+		status, _ := strconv.ParseInt(m[1], 10, 0)
+		message := m[2]
+		if status == http.StatusUnauthorized && (runner.currentDB.user == nil || runner.currentDB.user.Name() != "") {
+			status = http.StatusForbidden
+		}
+		return base.HTTPErrorf(int(status), "%s (while calling %s %q)", message, runner.kind, runner.name)
+	}
+	return &jsError{err, runner.kind, runner.name}
+}
+
+//////// DATABASE CALLBACK FUNCTION IMPLEMENTATIONS:
+
+// Implementation of JS `app.func(name, params)` function
+func (runner *userJSRunner) do_func(funcName string, params map[string]interface{}) (interface{}, error) {
+	return runner.currentDB.CallUserFunction(funcName, params, runner.mutationAllowed)
+}
+
+// Implementation of JS `app.get(docID, docType)` function
+func (runner *userJSRunner) do_get(docID string, docType *string) (interface{}, error) {
+	if err := runner.currentDB.CheckTimeout(); err != nil {
+		return nil, err
+	}
+	rev, err := runner.currentDB.GetRev(docID, "", false, nil)
+	if err != nil {
+		status, _ := base.ErrorAsHTTPStatus(err)
+		if status == http.StatusNotFound {
+			// Not-found is not an error; just return null.
+			return nil, nil
+		}
+		return nil, err
+	}
+	body, err := rev.Body()
+	if err != nil {
+		return nil, err
+	}
+	if docType != nil && body["type"] != *docType {
```

Comment:
`docType` is an optional parameter to the JS function:
- `get(id)` just gets the document with that docID.
- `get(id, docType)` gets the document but also verifies that its `type` property has the expected value.
I found this useful in implementing the travel-sample schema; for instance the implementation of the resolver for `airport(id:ID)` is `{return context.app.get(args.id, 'airport');}. That way if the caller passes the ID of a document that isn't an airport, the resolver fails instead of violating type-safety.

Reply:
I can see how this would be useful, was just wondering about decoupling it from the from the hardcoded "type" property for compatibility with existing datasets that use a different property to define type.  If this is primarily used for testing at the moment, can leave that as a future (possible) enhancement?

Reply:
I was under the impression that using `type` was very common ... if it's not, I can take this feature out.

Reply:
We can leave "but I use 'docType' instead of 'type'" as a problem to be addressed with a future enhancement as needed.